Source/Core/Common/Network.cpp

```diff
@@ -222,32 +227,50 @@ DHCPBody::DHCPBody(u32 transaction, MACAddress client_address, u32 new_ip, u32 s
   server_ip = serv_ip;
 }
 
-DHCPBody::DHCPBody(const std::vector<u8>& data)
+DHCPView::DHCPView() = default;
+
+DHCPView::DHCPView(const std::vector<u8>& data)
 {
-  if (data.size() > SIZE)
+  if (data.size() < DHCPBody::SIZE)
     return;
+  body = Common::BitCastPtr<DHCPBody>(data.data());
+  u16 opt_ptr = DHCPBody::SIZE;
+
+  while (opt_ptr < data.size())
+  {
+    const u8 fnc = Common::BitCastPtr<u8>(data.data() + opt_ptr);
+    const u8 len = Common::BitCastPtr<u8>(data.data() + opt_ptr + 1);
+    if (fnc == 255)
+      break;
+    const std::vector<u8> opt =
+        std::vector<u8>(data.data() + opt_ptr, data.data() + opt_ptr + 2 + len);
+    options.push_back(opt);
+    opt_ptr += 2 + len;
+  }
+}
 
-  std::memcpy(this, data.data(), data.size());
+void DHCPView::AddOption(u8 size, u8 fnc, const std::vector<u8>& params)
+{
+  std::vector<u8> opt = std::vector<u8>(params);
+  opt.insert(opt.begin(), size);
+  opt.insert(opt.begin(), fnc);
+  options.push_back(opt);
```

Comment:
Inserting element at the front of the vector might be costly. You can also get the size from the vector itself.
```C++
void DHCPView::AddOption(u8 fnc, const std::vector<u8>& params)
{
  std::vector<u8> opt = {fnc, u8(params.size())};
  opt.insert(opt.end(), params.begin(), params.end());
  options.emplace_back(std::move(opt));
}
```